llama : support RWKV v6 models

MollySophia opened the pull request:
This should fix #846.

## Added:
### ggml:
- Added unary operation ``Exp``
- Added ``rwkv_wkv`` operation with CPU impl
- Added ``rwkv_token_shift`` operation with CPU impl to handle multiple sequences in parallel(may not be necessary after #8526 is done)
### llama.cpp:
- ``rwkv_world`` tokenizer support (by @LaylBongers)
- ``convert_hf_to_gguf.py`` support for converting RWKV v6 HF models
- RWKV v6 graph building

## TODO:
- ~~Do modifications after #8526 is ready accordingly~~ Done
- ~~Add CUDA or Metal implementation for ``rwkv_wkv`` operation~~ Maybe next PR

##
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Medium

LaylBongers committed:
Add RWKV tokenization

LaylBongers committed:
Do not use special tokens when matching in RWKV tokenizer

LaylBongers committed:
Fix model loading

LaylBongers committed:
Add (broken) placeholder graph builder for RWKV

LaylBongers committed:
Add workaround for kv cache

LaylBongers committed:
Add logits conversion to rwkv5

LaylBongers committed:
Add rwkv5 layer norms

LaylBongers committed:
Add time mix KVRG & correct merge mistake

LaylBongers committed:
Add remaining time mix parameters

LaylBongers committed:
Add time mix output loading

LaylBongers committed:
Add placeholder llm_build_time_mix

MollySophia committed:
convert_hf_to_gguf: Add support for RWKV v6

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Fix build

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Fix build

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Load more tensors for rwkv v6

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Fix rwkv tokenizer

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
RWKV v6 graph building

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
ggml: Add unary operator Exp

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Add ``rescale_every_n_layers`` parameter

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Add ``wkv.head_size`` key for RWKV

so it doesn't reuse Mamba ssm parameters

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Fix offloading layers to CUDA

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Fix parallel inferencing for RWKV

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Remove trailing whitespaces

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -2702,6 +2702,80 @@ class StarCoder2Model(Model):
+            if new_name.endswith("time_mix_output.weight") or new_name.endswith("channel_mix_value.weight"):
+                data_torch = data_torch.div_(2 ** int(bid // rescale_every_n_layers))
+
+        tensors: list[tuple[str, Tensor]] = [(new_name, data_torch)]
+
+        return tensors
```
```suggestion
        yield (new_name, data_torch)
```

Still returning an `Iterable[tuple[str, Tensor]]`, but without having to make a list.

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -2702,6 +2702,80 @@ class StarCoder2Model(Model):
+        with open(self.dir_model / "rwkv_vocab_v20230424.txt", "r", encoding="utf-8") as f:
+            lines = f.readlines()
+            for line in lines:
+                x = eval(line[line.index(' '):line.rindex(' ')])
```
This can execute abritrary code, as long as it's in a file named `rwkv_vocab_v20230424.txt`.

Use [`ast.literal_eval`](https://docs.python.org/3/library/ast.html#ast.literal_eval) instead (by also adding `import ast` in this file) to avoid allowing to execute arbitrary expressions, while still parsing strings.

I think `line.split(' ')` could likely be used instead of using `line.index(' ')` and `line.rindex(' ')`.

Something like

```python3
parts = line.split(' ')
assert len(parts) >= 3
token_id, token, token_len = int(parts[0]), ast.literal_eval(' '.join(parts[1:-1])), int(parts[-1])

compilade reviewed (commented):
A few things I've noticed. I'll review this more deeply in the next days.

MollySophia commented on src/llama.cpp:
```diff
@@ -8357,6 +8559,208 @@ static struct ggml_tensor * llm_build_kv(
+    size_t n_kv = state_seq->ne[0];
+
+    struct ggml_tensor * sx = ggml_sub(ctx, x_prev, current);
+    struct ggml_tensor * xxx = ggml_add_inplace(
```
Didn't know about this. Thanks!

MollySophia committed:
build_rwkv: Avoid using inplace operations

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
convert_hf_to_gguf: rwkv: Avoid using ``eval``

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

Ronsor commented on ggml/src/ggml.c:
```diff
@@ -18047,6 +18388,8 @@ static void ggml_compute_backward(struct ggml_context * ctx, struct ggml_tensor
             } break;
```
`GGML_UNARY_OP_EXP` should probably be handled explicitly here. The backward computation is the same as the forward.

MollySophia replied:
```
case GGML_UNARY_OP_EXP:
    {
        if (src0->grad) {
            src0->grad = ggml_add_or_set(ctx,
                src0->grad,
                ggml_mul(ctx, tensor, tensor->grad),
                zero_table);
             }
    } break;
```
Like this?

Ronsor replied:
Looks correct to me 👍

Ronsor commented on convert_hf_to_gguf.py:
```diff
@@ -2702,6 +2703,80 @@ class StarCoder2Model(Model):
+                token_text: str = ""
+                for b in token:
+                    token_text += f"\\x{b:02x}"
+                tokens.append(token_text.encode("utf-8"))
```
Why are the token bytes converted to an escape sequence rather than saved directly?

MollySophia replied:
That's because the rwkv_world tokenizer contains incomplete UTF-8 bytes in decoded sequences. However, llama.cpp seems to require valid UTF-8 texts.

Ronsor replied:
Noted. The manual escaping could be avoided by doing `token_text = str(token)[2:-1]`.

```
>>> b = bytes("Testこんにちは\n123", "utf-8")
>>> b
b'Test\xe3\x81\x93\xe3\x82\x93\xe3\x81\xab\xe3\x81\xa1\xe3\x81\xaf\n123'
>>> str(b)
"b'Test\\xe3\\x81\\x93\\xe3\\x82\\x93\\xe3\\x81\\xab\\xe3\\x81\\xa1\\xe3\\x81\\xaf\\n123'"
>>> str(b)[2:-1]
'Test\\xe3\\x81\\x93\\xe3\\x82\\x93\\xe3\\x81\\xab\\xe3\\x81\\xa1\\xe3\\x81\\xaf\\n123'
>>> print(str(b)[2:-1])
Test\xe3\x81\x93\xe3\x82\x93\xe3\x81\xab\xe3\x81\xa1\xe3\x81\xaf\n123
```

As far as I know, `str(bytes_object)` will return a string with only normal ASCII characters, `\t`, `\n`, `\r`, or hex escape sequences.

MollySophia replied:
Tested and it looks good to me. Thanks for your help!

Ronsor commented on src/llama-vocab.cpp:
```diff
@@ -1080,6 +1080,109 @@ struct llm_tokenizer_ugm {
+
+struct llm_tokenizer_rwkv {
+    llm_tokenizer_rwkv(const llama_vocab & vocab): vocab(vocab) {
+        // RWKV supports arbitrary byte tokens, but the vocab struct only supports string tokens.
```
As a note, `std::string` can contain embedded `\0`s and other non-text characters, despite its name. I'm not sure if that was the concern.

Also, `naive_trie` could be used in here to improve performance. Since its methods take a `const char*` + a length as `size_t`, there shouldn't be any issue with `\0`s, I believe.

MollySophia replied:
The tokenizer implementation is done by @LaylBongers. I think she's working on optimizing this.

MollySophia committed:
convert_hf_to_gguf: rwkv tokenizer: Don't escape sequences manually

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia commented on convert_hf_to_gguf.py:
```diff
@@ -2702,6 +2703,80 @@ class StarCoder2Model(Model):
+            for line in lines:
+                parts = line.split(' ')
+                assert len(parts) >= 3
+                _, token, token_len = int(parts[0]), ast.literal_eval(' '.join(parts[1:-1])), int(parts[-1])
```
You are right XD

MollySophia committed:
Update convert_hf_to_gguf.py

Co-authored-by: compilade <git@compilade.net>

MollySophia committed:
ggml: Add backward computation for unary op ``exp``

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
Update convert_hf_to_gguf.py

Co-authored-by: compilade <git@compilade.net>

MollySophia committed:
Update convert_hf_to_gguf.py

Co-authored-by: compilade <git@compilade.net>

MollySophia commented on convert_hf_to_gguf.py:
```diff
@@ -2702,6 +2703,78 @@ class StarCoder2Model(Model):
     model_arch = gguf.MODEL_ARCH.STARCODER2
 
 
+@Model.register("Rwkv6ForCausalLM")
+class RwkvModel(Model):
+    model_arch = gguf.MODEL_ARCH.RWKV
```
This is a good point. When writing this part, I only thought about v6 support and possibly v5 (since it's pretty easy to keep the same arch if it's just for these two versions).
I guess it's better using MODEL_ARCH.RWKV6 here. Thanks!

MollySophia commented on src/llama.cpp:
```diff
@@ -8357,6 +8559,188 @@ static struct ggml_tensor * llm_build_kv(
+        xxx
+    );
+
+    struct ggml_tensor *mw = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, n_embed * n_tokens);
+    mw = ggml_reshape_2d(
+        ctx,
+        ggml_set_1d(ctx, mw, ggml_view_1d(ctx, xxx, n_embed * n_tokens, 0), 0),
+        n_embed, n_tokens
+    );
+
+    struct ggml_tensor *mk = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, n_embed * n_tokens);
+    mk = ggml_reshape_2d(
+        ctx,
+        ggml_set_1d(ctx, mk, ggml_view_1d(ctx, xxx, n_embed * n_tokens, n_embed * n_tokens * sizeof(float)), 0),
+        n_embed, n_tokens
+    );
+
+    struct ggml_tensor *mv = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, n_embed * n_tokens);
+    mv = ggml_reshape_2d(
+        ctx,
+        ggml_set_1d(ctx, mv, ggml_view_1d(ctx, xxx, n_embed * n_tokens, n_embed * n_tokens * 2 * sizeof(float)), 0),
+        n_embed, n_tokens
+    );
+
+    struct ggml_tensor *mr = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, n_embed * n_tokens);
+    mr = ggml_reshape_2d(
+        ctx,
+        ggml_set_1d(ctx, mr, ggml_view_1d(ctx, xxx, n_embed * n_tokens, n_embed * n_tokens * 3 * sizeof(float)), 0),
+        n_embed, n_tokens
+    );
+
+    struct ggml_tensor *mg = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, n_embed * n_tokens);
+    mg = ggml_reshape_2d(
+        ctx,
+        ggml_set_1d(ctx, mg, ggml_view_1d(ctx, xxx, n_embed * n_tokens, n_embed * n_tokens * 4 * sizeof(float)), 0),
+        n_embed, n_tokens
+    );
```
There isn't XD. It's just that I didn't know it can use ggml_view_2d here. I'll take a look at this.

MollySophia replied:
Done!

MollySophia committed:
Use MODEL_ARCH.RWKV6 instead of MODEL_ARCH.RWKV

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
build_rwkv6: Simplify graph

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Detect model.type

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Fix tensor loading for 7B/14B models

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia commented on src/llama.cpp:
```diff
@@ -8352,6 +8566,166 @@ static struct ggml_tensor * llm_build_kv(
+
+    // ggml_group_norm considers groups in the third dimension.
+    cur = ggml_reshape_4d(ctx, cur, 1, 1, n_embed, n_tokens);
+    cur = ggml_group_norm(ctx, cur, head_count, 64e-5f);
```
Found a problem here. Metal implementation of groupnorm requires ne0 % 4 == 0.

MollySophia replied:
Should be fixable with
```
cur = ggml_reshape_4d(ctx, cur, n_embed / head_count, 1, head_count, n_tokens);
cur = ggml_group_norm(ctx, cur, head_count, 64e-5f);
```

MollySophia committed:
llama: rwkv6: Fix group_norm assertion failure with Metal

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Clean up

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Add quantization tensor exclusion

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Use the new advanced batch splits

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia commented:
Synced the changes and made it working again after #8526 being merged.
This PR should be ready for review again now :D 
@compilade Could you take a look when convenient?

compilade commented on src/llama.cpp:
```diff
@@ -2151,6 +2216,7 @@ enum e_model {
     MODEL_1B,
     MODEL_1_3B,
     MODEL_1_4B,
+    MODEL_1_6B,
```
The string representation of that size should be handled too

```diff
diff --git a/src/llama.cpp b/src/llama.cpp
index 96725500..1aadab4c 100644
--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -5161,6 +5161,7 @@ static const char * llama_model_type_name(e_model type) {
         case MODEL_1B:            return "1B";
         case MODEL_1_3B:          return "1.3B";
         case MODEL_1_4B:          return "1.4B";
+        case MODEL_1_6B:          return "1.6B";
         case MODEL_2B:            return "2B";
         case MODEL_2_8B:          return "2.8B";
         case MODEL_3B:            return "3B";
```

MollySophia replied:
Yes XD. I missed that one

compilade reviewed (commented):
I'm impressed that `ggml_rwkv_wkv` only takes around 2% of the CPU time during inference of the 1.6B RWKV-v6 model (when measured with `perf record --call-graph=lbr`).

I have some styling comments, some suggestions, and I also found some problems.

MollySophia committed:
Update src/llama.cpp

Co-authored-by: compilade <git@compilade.net>

MollySophia commented:
> I'm impressed that `ggml_rwkv_wkv` only takes around 2% of the CPU time during inference of the 1.6B RWKV-v6 model (when measured with `perf record --call-graph=lbr`).
> 
> I have some styling comments, some suggestions, and I also found some problems.

Indeed. I did consider writing a metal kernel for wkv, but it turned out that wkv kernels didn't eat much cpu time.
I've also tried modfying current rwkv_wkv impl with GGML_SIMD macros, but the speed was almost the same. (Clang already did optimizations like vectorization, so writing manually may not be that necessary)

MollySophia committed:
llama: rwkv6: Use ``ggml_norm`` instead of ``ggml_group_norm``

Co-authored-by: compilade <git@compilade.net>

MollySophia commented on convert_hf_to_gguf.py:
```diff
@@ -2712,6 +2713,78 @@ class StarCoder2Model(Model):
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        new_name = self.map_tensor_name(name)
+
+        if not (new_name.endswith(".weight") or new_name.endswith(".bias")):
+            new_name += ".weight"
```
Somehow after changing ``or not name.endswith(".weight")`` with ``or not new_name.endswith(".weight")`` the converted gguf will make a CUDA error when inferencing. Gonna keep the existing one for now in this PR

compilade replied:
> Somehow after changing `or not name.endswith(".weight")` with `or not new_name.endswith(".weight")` the converted gguf will make a CUDA error when inferencing. Gonna keep the existing one for now in this PR

Models quantized with `llama-quantize` will *also* have this problem, since using `new_name` in the convert script simply makes it use the same types as `llama-quantize`. I think both methods should result in the same model files.

Which tensors are problematic? Maybe they need to be explicitly excluded from quantization (assuming they are small enough)?

MollySophia replied:
> > Somehow after changing `or not name.endswith(".weight")` with `or not new_name.endswith(".weight")` the converted gguf will make a CUDA error when inferencing. Gonna keep the existing one for now in this PR
> 
> Models quantized with `llama-quantize` will _also_ have this problem, since using `new_name` in the convert script simply makes it use the same types as `llama-quantize`. I think both methods should result in the same model files.
> 
> Which tensors are problematic? Maybe they need to be explicitly excluded from quantization (assuming they are small enough)?

Yes, they are not that large.
I've added the commit that matches new_name, as well as one that adds two problematic tensors in f32 list

MollySophia committed:
llama: rwkv6: Apply code style and misc changes

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia commented on src/llama.cpp:
```diff
@@ -9157,6 +9371,164 @@ static struct ggml_tensor * llm_build_mamba(
+
+    struct ggml_tensor * xk = ggml_add(
+        ctx,
+        ggml_mul(
+            ctx,
+            ggml_add(ctx, mk, layer->time_mix_lerp_k),
+            sx
+        ),
+        cur
+    );
+
+    struct ggml_tensor * xv = ggml_add(
+        ctx,
+        ggml_mul(
+            ctx,
+            ggml_add(ctx, mv, layer->time_mix_lerp_v),
+            sx
+        ),
+        cur
+    );
+
+    struct ggml_tensor * xr = ggml_add(
+        ctx,
+        ggml_mul(
+            ctx,
+            ggml_add(ctx, mr, layer->time_mix_lerp_r),
+            sx
+        ),
+        cur
+    );
+
+    struct ggml_tensor * xg = ggml_add(
+        ctx,
+        ggml_mul(
+            ctx,
+            ggml_add(ctx, mg, layer->time_mix_lerp_g),
+            sx
+        ),
+        cur
+    );
```
Current one resembles the model structure better.
I'm not sure if doing a single ``ggml_mul_mat`` before spliting would affect the performance, as well as which performs better.
Speaking of which, the first ``ggml_norm`` happening before all the layers can also be avoided by pre-calculating all the embedding vectors with a norm. I'm not sure if that should be done either

MollySophia committed:
converter: Use class name ``Rwkv6Model``

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Make use of key ``feed_forward_length``

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Add kv ``time_mix_extra_dim`` and ``time_decay_extra_dim``

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia commented on src/llama.cpp:
```diff
@@ -8199,6 +8350,69 @@ static bool llm_load_tensors(
+                    const int time_decay_extra_dim = (n_embd == 4096) ? 128 : 64;
+                    const int head_size = hparams.wkv_head_size;
+                    const int attn_hidden_size = n_embd;
+                    const int ffn_size = (int)(n_embd * 3.5 / 32) * 32;
```
I modified to use ``feed_forward_length`` for ``ffn_size`` here.
For the ``time_mix_extra_dim`` and ``time_decay_extra_dim``, I was thinking about keeping it like this but decided to add two keys for these eventually.
It's indeed better not to hardcode these so that things like breaking compatibility with already converted gguf files won't happen in the future.

MollySophia committed:
converter: Match ``new_name`` instead of ``name`` for float32 explicit tensors

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Keep ``time_mix_w1/w2`` as F32

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

compilade commented on src/llama.cpp:
```diff
@@ -14678,6 +15060,126 @@ struct llm_build_context {
+        cur = ggml_get_rows(ctx0, cur, inp_out_ids);
+
+        cur = llm_build_norm(ctx0, cur, hparams, model.output_norm, model.output_norm_b, LLM_NORM, cb, -1);
+        cur = ggml_mul_mat(ctx0, model.output, cur);
```
Are there any LoRA adapters for RWKV-v6? If it's desirable to support them, then most `ggml_mul_mat` can be replaced with `llm_build_lora_mm`.

See #8332 for some background if you want.

```suggestion
        cur = llm_build_lora_mm(lctx, ctx0, model.output, cur);
```

MollySophia replied:
I'll take a look at LoRA things for RWKV v6 this week.
I tend to have the base support merged first lol. I'll make a new PR for LoRA support later.

MollySophia replied:
I've added llm_build_lora_mm to some supported matmuls according to https://github.com/JL-er/RWKV-PEFT
Specifically head.weight, att.key/receptance/value/gate/output; ffn.key/value/receptance.

BTW What's next for this PR?

MollySophia commented on src/llama.cpp:
```diff
@@ -14678,6 +15060,126 @@ struct llm_build_context {
+            cur = ggml_add(ctx0, cur, llm_build_channel_mix_rwkv6(ctx0, layer, x_norm_ffn, x_prev));
+            ggml_build_forward_expand(gf, cur);
+
+            token_shift = ggml_cont(
+                ctx0,
+                ggml_permute(
+                    ctx0,
+                    ggml_reshape_3d(ctx0, token_shift, n_embd, n_seqs, 2),
+                    0, 2, 1, 3
+                )
+            );
```
I guess yes lol

MollySophia committed:
llama: rwkv6: Remove unused nodes

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Apply code format changes

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia committed:
llama: rwkv6: Add lora for some supported tensors

Currently att.key/receptance/value/gate/output, ffn.receptance/key/value, as well as head.weight

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

ggerganov commented:
Lets look to merge soon. @MollySophia Which HF model do you recommend to run a few tests with this branch?

MollySophia commented:
> Lets look to merge soon. @MollySophia Which HF model do you recommend to run a few tests with this branch?

https://huggingface.co/RWKV/v6-Finch-1B6-HF should be enough for testing the functionalities.
https://huggingface.co/RWKV/v6-Finch-7B-HF/tree/main or the 3B one should be working too

ggerganov committed:
rwkv : speed-up tokenization using trie

ggerganov commented:
I've updated the tokenizer to use a true for string search (7004323). With this change the time for tokenizing `wiki.test` dropped from 27s to 40ms on my Mac.

ggerganov committed:
minor : style + indentation

compilade commented on src/llama.cpp:
```diff
@@ -14680,6 +15069,117 @@ struct llm_build_context {
+                )
+            );
+
+            if ((il + 1) % hparams.rescale_every_n_layers == 0) {
```
The default value for `hparams.rescale_every_n_layers` is `0` and it's marked optional when loading with `ml.get_key` initially, which would cause a division by zero. Either `0` needs to be specially handled in the model graph (like suggested here), or it should not be optional when loading.

```suggestion
            if (hparams.rescale_every_n_layers != 0 && (il + 1) % hparams.rescale_every_n_layers == 0) {
```

MollySophia replied:
Indeed. Thanks!

compilade reviewed (approved):
> BTW What's next for this PR?

@MollySophia It looks ready for me, at least. Nice work!

There's some potential division by zero with `hparams.rescale_every_n_layers` which I think should be fixed before merging.

Improvements to `ggml_rwkv_wkv` (if relevant) can be done later in a follow-up PR, so I think this will be ready to merge.

MollySophia commented on ggml/src/ggml.c:
```diff
@@ -16734,6 +16850,95 @@ static void ggml_compute_forward_add_rel_pos(
+        return;
+    }
+
+    memset(dst_data, 0, T * C * sizeof(float));
```
I think yes. Commenting out the memset makes the outputs insane :D

ggerganov replied:
Ideally, it should not be necessary, but the current implementation accumulates into existing values:

```c
dst_data[t_h_j_offset] += temp_val * r_val;
```

So these have to be initialized to 0 with the `memest`. We can think of ways to improve this later.

MollySophia committed:
llama: rwkv6: Avoid division by zero

Co-authored-by: compilade <git@compilade.net>

MollySophia committed:
ggml: rwkv_wkv: Avoid copying the state

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

MollySophia commented on ggml/src/ggml.c:
```diff
@@ -16734,6 +16850,95 @@ static void ggml_compute_forward_add_rel_pos(
+    float * r =          (float *) dst->src[2]->data;
+    float * time_faaaa = (float *) dst->src[3]->data;
+    float * time_decay = (float *) dst->src[4]->data;
+    memcpy(state, dst->src[5]->data, (C / H) * C * n_seqs * sizeof(float));
```
Done that since it's an easy fix anyway :)
